### PR TITLE
Improve emoteevBidAdapter

### PR DESCRIPTION
**Pull request by piotr-yuxuan**
## Type of change

- [ ] Bugfix
- [x] Feature
- [ ] New bidder adapter
- [x] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] This change _doesn't_ affect user-facing APIs or examples documented on http://prebid.org.
- [ ] Other

#### Squash several pending changes

Following the long-awaited PR https://github.com/prebid/Prebid.js/pull/3391, here are some changes we can now release. We avoided to modify too much the previous PR to keep things easy for the reviewer.

- Improve test coverage
- Document important constants
- Extreme programming: document all functions
- Imperative shell, functional core
- Send events `onAdapterCalled`, `onBidWon`, and `onTimeout`
- Report GDPR relevance and consent

Please note that some event pixels might trigger a `422 Unprocessable Entity` response from our side. This doesn't preclude bid requests to be responded to, or won bid to be displayed.

#### Test coverage

Tests pass with HeadlessChrome 73.0.3683 (Mac OS X 10.14.3), including webGL tests. ? However IE and Edge don't support these tests (cf. 2679693...ed0b9ac) so commenting them out to make this PR pass CI. As a result, downgrading test coverage to:

- 94.5% Statements 103/109
- 98.18% Branches 54/55
- 100% Functions 28/28
- 93.88% Lines 92/98

#### Integration tests

Tested against production endpoint with the possible combinations of these following parameters:

- Browser:
  · Chrome Canary 75.0.3739.0
  · Firefox Developer Edition 67.0b3 (64-bit)
  · Safari version 12.0.3 (14606.4.5)
  · Tor Browser 8.0.6 (based on Mozilla Firefox 60.5.1esr)
- Integration page:
  · integrationExamples/gpt/hello_world_emoteev.html
  · https://jsfiddle.net/8aqotw1k/6/

Test parameters for validating bids:

```
{
  bidder: 'emoteev',
  params: {
    adSpaceId: 5084
  }
}
```

Localhost test server launched with:
```
$ npm install && gulp serve
```

## Contact

- @jeremyvdw, owner and gate keeper of this code
- @piotr-yuxuan, owner of this code

### CODE REVIEW

**Comment by piotr-yuxuan** [comment]
Has anybody an idea to make webGL tests pass on IE and Edge? See 2679693...ed0b9ac for faulty tests.

File: modules/emoteevBidAdapter.js [hunk 1]
```diff
@@ -153,62 +422,70 @@ export const getDeviceInfo = (deviceDimensions, viewDimensions, documentDimensio
...
-    }
-    if (syncOptions.pixelEnabled && serverResponses.length > 0) {
-      syncs.push({
-        type: 'image',
-        url: userSyncImageUrl(parameteremoteevEnv, configemoteevEnv),
-      });
+  isBidRequestValid: (bidRequest) => {
+    const isValid = isBidRequestValid(bidRequest);
+    if (isValid) {
+      triggerPixel(url.format(onAdapterCalled(
```
Comment (mike-chowla) [inline modules/emoteevBidAdapter.js]: Adapters are prohibited from firing their own pixels.  If you want fire pixels , you can return them from get getUserSyncs

File: modules/emoteevBidAdapter.js [hunk 2]
```diff
@@ -153,62 +422,70 @@ export const getDeviceInfo = (deviceDimensions, viewDimensions, documentDimensio
...
+  buildRequests: (validBidRequests, bidderRequest) =>
+    buildRequests(
+      resolveEnv(config.getConfig(), getParameterByName('emoteevEnv')),
+      resolveDebug(config.getConfig(), getParameterByName('debug')),
+      config.getConfig('currency'),
+      validBidRequests,
+      bidderRequest),
+  interpretResponse: interpretResponse,
+  onBidWon: (bidObject) =>
+    triggerPixel(url.format(onBidWon(
```
Comment (mike-chowla) [inline modules/emoteevBidAdapter.js]: Adapters are prohibited from firing their own pixels.  If you want fire pixels , you can return them from get getUserSyncs

File: modules/emoteevBidAdapter.js [hunk 3]
```diff
@@ -153,62 +422,70 @@ export const getDeviceInfo = (deviceDimensions, viewDimensions, documentDimensio
...
+      resolveEnv(config.getConfig(), getParameterByName('emoteevEnv')),
+      getCookie('_pubcid'),
+      bidObject))),
+  onTimeout: (bidRequest) =>
+    triggerPixel(url.format(onTimeout(
```
Comment (mike-chowla) [inline modules/emoteevBidAdapter.js]: Adapters are prohibited from firing their own pixels.  If you want fire pixels , you can return them from get getUserSyncs

**Comment by piotr-yuxuan** [comment]
Hi @mike-chowla, thanks for you swift review!

We did chose to use `triggerPixel` because it looked more disciplined, and more evolved than manipulating raw `img.src`. Moreover we did chose to trigger pixels out of UserSyncs (namely in `onTimeout`, `onBidWon`, and `isBidRequestValid`) after reading through adapters code doing the same, so we do feel it's fair and right to do so.

These pixels would merely send us datapoints needed to monitor our adapter behaviour across the internet. The intent here is to get metrics as per the principles of [three pillars of observability](https://www.oreilly.com/library/view/distributed-systems-observability/9781492033431/ch04.html). I say it in good faith, we don't do that to track users, fiddle with their cookies, or doing nasty things. By the way, as suggested by [these lines](https://github.com/Prebid/Prebid.js/blob/26796938122ce3726343d54f43a4967b7a7ccb33/modules/emoteevBidAdapter.js#L308-L309) we are planning to be GDPR compliant as fast as possible.

The rest of this message will examplify how we intend to use data from these pixels to improve our adapter observability, then we skim through merged, release code and find some examples of code similar to our intended use.

> Because it's currently something accepted from other reviewers with no problem, and because it's currently used for fair purposes in other adapters, might I ask the tech team of Prebid.js to consider a second review?

### Improve `emoteevBidAdapter` observability

#### `on_adapter_called`

- This event would allow us to keep an eye on the average time between when `isBidRequestValid` returns true so that we can monitor our partner publishers integrations and help them correctly set up their ad units. As I read our code again I acknowledge we should actually send an event every time `isBidRequestValid` is called so we can even give our partner websites the percentage of valid request.
- Another interesting metric to observe is the time difference between `on_adapter_called` and our bid endpoint actually get the matching request. This could be helpful to the publishers to figure out if they call too much adapters, so that auctions become heavy and at the end they loose inventory.

#### `on_bid_won`

- This event enables us to monitor the discrepencies between bid actually won and creative actually displayed.
- Should a publisher grant us a bid, we want to make absolutely sure we are aware of it so we can pay the publisher transparently. An adapter pixel is a clean way of doing so. It's much more reliable than a pixel baked in the creative itself. Also, with this pixel we can efficiently spot problems in creative.
- We can monitor and compare the bids we've won against the creative we've chosen and the actually displayed creative. We want to live up to our commitment to spend efficiently each and every dollar an advertiser entrusts us with.
- Finally it would enable us to provide fair metrics to our partners on the advertising side to honestly give them our win rate.

#### `on_timeout`

- When a publisher activates our adapter on their site, they want us to reply and send back bids to their requests. This event enables us to make sure we keep being fast enough for them so we get things done under the timeout they set.
- Also, should a publisher makes a mistake and set a value too low, this would make it possible for us to debug and pro-actively advice them to reconsider their settings.

### Current pixels outside of UserSyncs

_Mgid_ [triggers](https://github.com/Prebid/Prebid.js/blob/849ae8f092268823a2cc98cca58056451ac47518/modules/mgidBidAdapter.js#L221) a raw pixel in `onBidWon`. This [commit](https://github.com/Prebid/Prebid.js/commit/849ae8f092268823a2cc98cca58056451ac47518) has been created on 22 Fev 2019 and [approved](https://github.com/prebid/Prebid.js/pull/3562#event-2156885472) the same day by @idettman. It's just a couple of weeks ago. It looks like the current usage.

_Zedo_ uses `utils.triggerPixels` to fire pixels in [`onTimeout`](https://github.com/Prebid/Prebid.js/blob/5ecf54b0a981f440023d83ff443f709dff4f8334/modules/zedoBidAdapter.js#L148) and [`onBidWon`](https://github.com/Prebid/Prebid.js/blob/5ecf54b0a981f440023d83ff443f709dff4f8334/modules/zedoBidAdapter.js#L156). These changes got in PR https://github.com/prebid/Prebid.js/pull/3240 and where [approved](https://github.com/prebid/Prebid.js/pull/3240#pullrequestreview-172054400) by @jaiminpanchal27 on 6 Nov 2018, a mere four months ago.

_Medianet_ uses `utils.triggerPixels` to fire pixels in [`onTimeout`](https://github.com/Prebid/Prebid.js/blob/2c04c0168b14ec6c6f5a28b831b95c71a46907b5/modules/medianetBidAdapter.js#L367) and [`onBidWon`](https://github.com/Prebid/Prebid.js/blob/2c04c0168b14ec6c6f5a28b831b95c71a46907b5/modules/medianetBidAdapter.js#L380). These changes got in PR https://github.com/prebid/Prebid.js/pull/2954 and where [approved](https://github.com/prebid/Prebid.js/pull/2954#pullrequestreview-154359026) by @harpere on 11 Sep 2018, that's to say six months ago.

Moreover _JustPremium_ uses raw `img.src` to fire pixels in [`onTimeout`](https://github.com/Prebid/Prebid.js/blob/8635c91942de9df4ec236672c39b19448545a812/modules/justpremiumBidAdapter.js#L118) and this was [approved](https://github.com/prebid/Prebid.js/pull/3238#event-1935424718) by @jsnellbaker on 30 Oct 2018 with no remarks.

_Mantis_ uses raw `img.src` to [fire pixels](https://github.com/Prebid/Prebid.js/commit/f7c7905638a37b478b85f0e64971d3d24ddb55fe) on master since it's been [approved](https://github.com/prebid/Prebid.js/pull/2696#pullrequestreview-127375319) on 9 Jun 2018 by @idettman.

**Comment by piotr-yuxuan** [comment]
Hi @mike-chowla, thanks for your swift review!

We did choose to use `triggerPixel` because it looked more disciplined, and more evolved than manipulating raw `img.src`. Moreover we did choose to trigger pixels out of UserSyncs (namely in `onTimeout`, `onBidWon`, and `isBidRequestValid`) after reading through adapters code doing the same, so we do feel it's fair and right to do so.

These pixels would merely send us datapoints needed to monitor our adapter behaviour across the internet. The intent here is to get metrics as per the principles of [three pillars of observability](https://www.oreilly.com/library/view/distributed-systems-observability/9781492033431/ch04.html). I say it in good faith, we don't do that to track users, fiddle with their cookies, or to do nasty things. By the way, as suggested by [these lines](https://github.com/Prebid/Prebid.js/blob/26796938122ce3726343d54f43a4967b7a7ccb33/modules/emoteevBidAdapter.js#L308-L309) we are planning to be GDPR compliant as fast as possible.

### TL;DR

The rest of this message examplifies how we intend to use data from these pixels to improve our adapter observability, then we look at _what the documentation says_ about pixels, and finally we skim through merged, released code and find some examples of code similar to our intended use.

> Because it's currently something accepted from other reviewers with no problem, and because it's currently used for fair purposes in other adapters, might I ask the team of Prebid.js to contemplate a second review?

### Improve `emoteevBidAdapter` observability

#### `on_adapter_called`

- This event would allow us to keep an eye on the average time between when `isBidRequestValid` returns true and when we actually display a creative. As I read our code again I acknowledge we should actually send an event every time `isBidRequestValid` is called so we can even give our partner websites the percentage of valid request and report misconfigured ad units.
- Another interesting metric to observe is the time difference between `on_adapter_called` and our bid endpoint actually get the matching request. This could be helpful to the publishers to figure out if they call too much adapters, so that auctions become heavy and at the end they loose inventory.

#### `on_bid_won`

- This event enables us to monitor the discrepencies between bid actually won and creative actually displayed.
- Should a publisher grant us a bid, we want to make absolutely sure we are aware of it so we can pay the publisher transparently. An adapter pixel is a clean way of doing so. It's much more reliable than a pixel baked in the creative itself. Also, with this event we can efficiently spot problems in creative without penalizing the publisher.
- We can monitor and compare the bids we've won against the creative we've chosen and the actually displayed creative. We want to live up to our commitment to spend efficiently each and every dollar an advertiser entrusts us with.
- Finally it would enable us to provide fair metrics to our partners on the advertising side to honestly give them our win rate.

#### `on_timeout`

- When a publisher activates our adapter on their site, they want us to reply and send back bids to their requests. This event enables us to make sure we keep being fast enough for them so we get things done under the timeout they set.
- Also, should a publisher make a mistake and set a value too low, this would make it possible for us to debug and pro-actively advice them to reconsider their settings.

### Adapter documentation

On 16 Apr 2018 @mkendall07 [approved](https://github.com/prebid/prebid.github.io/pull/678#event-1577300890) PR https://github.com/prebid/prebid.github.io/pull/678 on the documentation about changes to `onTimeout` made by @vedantseta. It's still live, and be be read in the [current documentation](http://prebid.org/dev-docs/bidder-adaptor.html#registering-on-timeout):

> The onTimeout function will be called when an adpater timed out for an auction. Adapter can fire a ajax or pixel call to register a timeout at thier end.

In my humble opinion, the doc here shows a good idea: reacting to an event (timeout) with a side-effect (pixel call). We merely want to implement this idea and react to a timeout, but also a handful of other events: when our adapter is called, and when a bid is won.

### Current pixels outside of UserSyncs

Our current review says: « adapters are prohibited from firing their own pixels. If you want fire pixels, you can return them from get getUserSyncs ». However there are a few examples of recently approved and merged code which fire pixels out of UserSyncs. So we feel it's fair to do the same and gather data to monitor and improve our adapter.

_Mgid_ [triggers](https://github.com/Prebid/Prebid.js/blob/849ae8f092268823a2cc98cca58056451ac47518/modules/mgidBidAdapter.js#L221) a raw pixel in `onBidWon`. This [commit](https://github.com/Prebid/Prebid.js/commit/849ae8f092268823a2cc98cca58056451ac47518) has been created on 22 Fev 2019 and [approved](https://github.com/prebid/Prebid.js/pull/3562#event-2156885472) the same day by @idettman. It's just a couple of weeks ago. It looks like the current usage.

_Zedo_ uses `utils.triggerPixels` to fire pixels in [`onTimeout`](https://github.com/Prebid/Prebid.js/blob/5ecf54b0a981f440023d83ff443f709dff4f8334/modules/zedoBidAdapter.js#L148) and [`onBidWon`](https://github.com/Prebid/Prebid.js/blob/5ecf54b0a981f440023d83ff443f709dff4f8334/modules/zedoBidAdapter.js#L156). These changes got in PR https://github.com/prebid/Prebid.js/pull/3240 and where [approved](https://github.com/prebid/Prebid.js/pull/3240#pullrequestreview-172054400) by @jaiminpanchal27 on 6 Nov 2018, a mere four months ago.

_Medianet_ uses `utils.triggerPixels` to fire pixels in [`onTimeout`](https://github.com/Prebid/Prebid.js/blob/2c04c0168b14ec6c6f5a28b831b95c71a46907b5/modules/medianetBidAdapter.js#L367) and [`onBidWon`](https://github.com/Prebid/Prebid.js/blob/2c04c0168b14ec6c6f5a28b831b95c71a46907b5/modules/medianetBidAdapter.js#L380). These changes got in PR https://github.com/prebid/Prebid.js/pull/2954 and where [approved](https://github.com/prebid/Prebid.js/pull/2954#pullrequestreview-154359026) by @harpere on 11 Sep 2018, that's to say six months ago.

Moreover _JustPremium_ uses raw `img.src` to fire pixels in [`onTimeout`](https://github.com/Prebid/Prebid.js/blob/8635c91942de9df4ec236672c39b19448545a812/modules/justpremiumBidAdapter.js#L118) and this was [approved](https://github.com/prebid/Prebid.js/pull/3238#event-1935424718) by @jsnellbaker on 30 Oct 2018 with no remarks.

_Mantis_ uses raw `img.src` to [fire pixels](https://github.com/Prebid/Prebid.js/commit/f7c7905638a37b478b85f0e64971d3d24ddb55fe) on master since it's been [approved](https://github.com/prebid/Prebid.js/pull/2696#pullrequestreview-127375319) on 9 Jun 2018 by @idettman.

**Comment by mike-chowla** [comment]
Looks like some pixels have slipped when they should not have been approved.  Prebid.js' policy doesn't allow them.  The only way I can allow them is if @mkendall07 OKs them.

**Comment by piotr-yuxuan** [comment]
Hi @mike-chowla,

Thanks for your reply. In my humble opinion, the paragraph _Adapter documentation_ of my previous message might suggest pixels are actually allowed in event handlers like `onTimeout`. However I gonna await @mkendall07's response 🙂

Cheers

**Comment by jeremyvdw** [comment]
@mkendall07 Do you mind having a look at this? We are kind of stuck with this PR and it's now getting pretty urgent. Thank you

File: integrationExamples/gpt/pbjs_example_gpt.html [hunk 4]
```diff
@@ -314,7 +314,7 @@
                           height: '250',
                       }
                     },
-		            {
+                    {
```
Comment (mkendall07) [inline integrationExamples/gpt/pbjs_example_gpt.html]: please drop changes to this file.

File: modules/emoteevBidAdapter.js (hunk 1 above)
Comment (mkendall07) [reply modules/emoteevBidAdapter.js]: this pixel needs to be removed.

File: modules/emoteevBidAdapter.js (hunk 2 above)
Comment (mkendall07) [reply modules/emoteevBidAdapter.js]: this pixel is ok since it's tied to the event handler.

File: modules/emoteevBidAdapter.js (hunk 3 above)
Comment (mkendall07) [reply modules/emoteevBidAdapter.js]: this pixel is ok since it's tied to the event handler.

**Comment by mkendall07** [comment]
@piotr-yuxuan @jeremyvdw 
I've reviewed and put a few comments. We'll allow the pixels from the event handlers since it was previously allowed. Thanks.

**Comment by piotr-yuxuan** [comment]
@mkendall07 @mike-chowla thanks for your patience and your reviews. To the best of my knowledge this code should now follow your reviews guidelines and follow Prebid.js pixel policy. Please let me know any further needed changes.
